presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java

```diff
@@ -1507,6 +1509,161 @@ private static void validateSphericalType(String function, OGCGeometry geometry,
         }
     }
 
+    @SqlNullable
+    @Description("Returns the area of a geometry on the Earth's surface")
+    @ScalarFunction("ST_Area")
+    @SqlType(DOUBLE)
+    public static Double stSphericalArea(@SqlType(SPHERICAL_GEOGRAPHY_TYPE_NAME) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        if (geometry.isEmpty()) {
+            return null;
+        }
+
+        validateSphericalType("ST_Area", geometry, EnumSet.of(POLYGON, MULTI_POLYGON));
+
+        return sphericalPolygonArea((Polygon) geometry.getEsriGeometry());
+    }
+
+    private static double sphericalPolygonArea(Polygon polygon)
+    {
+        // See https://www.movable-type.co.uk/scripts/latlong.html
+        // and http://osgeo-org.1560.x6.nabble.com/Area-of-a-spherical-polygon-td3841625.html
+        // and https://www.element84.com/blog/determining-if-a-spherical-polygon-contains-a-pole
+        // for the underlying Maths
+
+        double sphericalExcess = 0.0;
+
+        int numPaths = polygon.getPathCount();
+        for (int i = 0; i < numPaths; i++) {
+            double sign = polygon.isExteriorRing(i) ? 1.0 : -1.0;
+            sphericalExcess += sign * sphericalPathExcess(polygon, i);
+        }
+
+        return Math.abs(sphericalExcess * EARTH_RADIUS_M * EARTH_RADIUS_M);
+    }
+
+    private static double sphericalPathExcess(Polygon polygon, int pathIndex)
+    {
+        int start = polygon.getPathStart(pathIndex);
+        int end = polygon.getPathEnd(pathIndex);
+        if (end - start < 3) {
+            // A path with less than 3 points is not valid for calculating an area
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("Can't calculate the area of a polygon with rings with less than 3 vertices. Input has %d", end - start));
+        }
+
+        // If our path is already closed, let's consider it to be shorter by 1 point
+        if (polygon.getPoint(start).equals(polygon.getPoint(end - 1))) {
+            end = end - 1;
+        }
+
+        // .getPoint allocates a new Point object on every call, unless we give it a point to store the result
+        Point tempOutPoint = new Point();
+
+        // In our path, the end point precedes the first one
+        polygon.getPoint(end - 1, tempOutPoint);
+        BearingAndExcessCalculator calculator = new BearingAndExcessCalculator(tempOutPoint);
+
+        for (int i = start; i < end; i++) {
+            polygon.getPoint(i, tempOutPoint);
+            calculator.processPoint(tempOutPoint);
+        }
+
+        return calculator.getSphericalExcess();
+    }
+
+    private static class BearingAndExcessCalculator
```

Comment:
Rename to SphericalExcessCalculator, as it produces only spherical excess.

Reply:
It does calculate the Bearing, even if it's not exposed (yet) , and it will probably be useful to determine a point outside the polygon (ie see if the pole is outside) which will be required to determine if a point is inside or outside a polygon ... So while I agree with you right now, I also don't want people to have to re-implement it later because it's less explicit that it does that as well. What do you think ?